packages/core/src/types.ts

```diff
@@ -75,6 +81,365 @@ export interface PositionWithOffset extends Position {
  */
 export type SourceRange = [number, number];
 
+//------------------------------------------------------------------------------
+// Rules
+//------------------------------------------------------------------------------
+
+/**
+ * What the rule is responsible for finding:
+ * - `problem` means the rule has noticed a potential error.
+ * - `suggestion` means the rule suggests an alternate or better approach.
+ * - `layout` means the rule is looking at spacing, indentation, etc.
+ */
+export type RuleType = "problem" | "suggestion" | "layout";
+
+/**
+ * The type of fix the rule can provide:
+ * - `code` means the rule can fix syntax.
+ * - `whitespace` means the rule can fix spacing and indentation.
+ */
+export type RuleFixType = "code" | "whitespace";
+
+/**
+ * An object containing visitor information for a rule. Each method is either the
+ * name of a node type or a selector, or is a method that will be called at specific
+ * times during the traversal.
+ */
+export interface RuleVisitor {
+	/**
+	 * Called for each node in the AST or at specific times during the traversal.
+	 */
+	[key: string]:
+		| ((node: unknown, parent?: unknown) => void)
+		| ((...unknown: unknown[]) => void);
+}
```

Comment:
How will this interface be used? Will it be extended to declare language-specific interfaces where `node` and `parent` have a particular type, or will rules simply create and return a `RuleVisitor` object?

Reply:
It's designed to be extended but still work as a default. (see the `RuleDefinition` interface)

Reply:
I tried to declare an interface that extends `RuleVisitor` with something more specific, but it's not clear to me how the TypeScript syntax would look like ([**Playground link**](https://www.typescriptlang.org/play/?#code/KYDwDg9gTgLgBASwHY2FAZgQwMbDgJQFcAbYANQQGcEZo4BvAKAEgB6AKnZbnbgGFMxUgBM46OsBwALOEgjC8yODCl4AggGUAKnDqZ4lMMGwJ0CbMoQBbYJTjDCUZAHNlq5VEwA3NJUEA6bnZWFgBtAGtgAE8ALjhKGCckZwBdGJZmAB84AAocuQU4wiRwuQB3JAAaODBMKGAUAH4ikvKkAEo4AF4APjgvCARhdozsvP8J4tKICpbpitCUzt7+weGAbkYAX0ZGGCijOABZKIA5eTwuhmUD4DiEpOd1-sFCO7gptrgtzcZkVAwODwJyIpAo1FoUDgoFQSGEdlB5CoNDoTGYEWi90SLjSuQK7xO5wU1Vq9SacUJF2WfQGQ02OyAA)).

Reply:
A possible solution: we could use generics in place of the actual argument types, e.g.:

```ts
export interface RuleVisitor<NodeType = unknown, ArgsType extends unknown[] = unknown[]> {
	[key: string]:
		| ((node: NodeType, parent?: NodeType) => void)
		| ((...unknown: ArgsType) => void);
}
```

Implementors could specify those types explicitly:

```ts
interface MyRuleVisitor extends RuleVisitor<MyNode> {
	[key: string]: (node: MyNode, parent?: MyNode) => void;
}
```

or

```ts
interface MyRuleVisitor extends RuleVisitor<never, [string, number, boolean]> {
	[key: string]: (arg1: string, arg2: number, arg3: boolean) => void;
}
```

[**Playground link**](https://www.typescriptlang.org/play/?#code/KYDwDg9gTgLgBASwHY2FAZgQwMbDgJQFcAbYANQQGcEZoAeAOQgBNgAVATzDwF45CkAayQQA7kgA0cAIJQA5pU7c4oVEmaV+QkeIDaAXTh8BwsUgMA+OAG8AUAEgA9ACpnDuM7gBhTMVLM4dGgVHAALOBFWRCQ4GFC8aQBlNjhgzHhKbmwEdARsWIQAW2BNZkIoZDlY+NioTAA3NEpfADp3Z0cHXUFgDgAuOEoYCqQ5fT6HewAfOAAKWcjgAaZWJWApMEwoYBQAfmWWdi5gAEojK3qIBGYTyZn5lseTHSQB2QU1s54Lq5uAblsAF9bLYYMc4ABZDgrXg2WLHAZDEZyP5weq+QhLLSmcRwQEA2zIVAYHB4KFEUgUai0KAqEBqDQEEjkKg0ehQmFWOz2bq9RHDSrjOaLAYcw4bLY7GD7SHQw5fH7XAHAwkoNBYXCyiks6nQABMdIZmm1VLZUDoSGAjSgUl0SMqUiQhEKACM0FIXRAIKRMEh9Fyuj1+oMBaMhbMtnIAIz85FSSN6gZO13uuCRgDMA093uAvoVaN+ytsQA)

Reply:
Is there any value in having an empty interface? Isn't that just like putting `any` (or `unknown`)?

I was looking for something where everyone wouldn't need to create an index signature for their own rule visitor. For methods with esquery selectors, they'd need to override `[key: string]` to get a valid type for that kind of method.

Reply:
It's true, an empty interface can be assigned any value when the `strictNullChecks` TypeScript option is not set, so it would be only useful as an indicator for people reading the code.

Another option is keeping the index signature with a generic function type, for example:

```ts
export interface RuleVisitor {
  [key: string]: (...args: any) => void;
}
```

This would still not prevent implementors from adding methods that return non-void types, or from adding any other properties with arbitrary types.

Reply:
Went with this.